helix-vcs/src/lib.rs

```diff
@@ -0,0 +1,47 @@
+use std::{collections::HashMap, path::Path};
+
+pub use differ::Differ;
+pub use git::Git;
+
+mod differ;
+mod git;
+mod rope_line_cache;
+
+// TODO: Move to helix_core once we have a generic diff mode
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum LineDiff {
+    Added,
+    Deleted,
+    Modified,
+}
+
+/// Maps line numbers to changes
+pub type LineDiffs = HashMap<usize, LineDiff>;
+
+trait DiffProvider {
+    fn get_file_head(&self, file: &Path) -> Option<Vec<u8>>;
+}
+pub struct DiffProviderRegistry {
+    providers: Vec<Box<dyn DiffProvider>>,
+}
+
+impl DiffProviderRegistry {
+    pub fn new() -> DiffProviderRegistry {
+        // currently only git is supported
+        // TODO make this configurable when more providers are added
+        let git: Box<dyn DiffProvider> = Box::new(Git);
+        let providers = vec![git];
+        DiffProviderRegistry { providers }
```

Comment:
I don't think we need dynamic dispatch for something which can have all the known implementations implemented here, we can just use an enum to represent all `DiffProvider` to remove that `dyn` which seemed unnecessary.

Reply:
For this PR you are right. However I think dynamic dyspatch will be required in the future. There are many different VCS that people would like to use with helix (mecurial and pejul have been requested already). Adding all of these to core is not sustainable so they will either end up using a potential future plugin system or some other mechanism for dynamic loading (I have some ideas but really don't want to open that can of worms in this PR ). Furthermore I will open a followup PR once thisone is merged that contains a cache for repositories, That means that the Git Diff provider will have a state of some sort and probably be wrapped in an `Arc`/`Rc` so the price of boxing will have to be payed anyway

Reply:
True, sounds reasonable to me.